renamed module tls client auth params to avoid overlaping with fetch_url

coreaut opened the pull request:
##### SUMMARY
renamed the module parameters:
```
ca_cert
client_key
client_cert
```

because they wrongly overlapped with fetch_url allowed parameters.


Fixes #59513

##### ISSUE TYPE
- Bugfix Pull Request


##### COMPONENT NAME
pulp_repo

##### ADDITIONAL INFORMATION
To use a remote repo in pulp for sync which needs tls client auth (redhat repos) you need
to set this vars.
But they are also used in the ansible fetch_url function. So the module uses tls client auth also for API calls to the pulp server which is wrong.

It should only use tls client auth for the remote repo you wan't to sync. The Doku of the module says those vars are **only** for the remote repo.

Commit:
renamed module tls client auth parameters to avoid overlaping with ansible fetch_url

Commit:
added version_added info for params

coreaut commented:
To better understand why the parameter names are wrong:

https://github.com/ansible/ansible/issues/59513#issuecomment-514631983

They have accidentally been renamed in #54315

sysadmind reviewed (approved):
Nice legwork on this one. I haven't used pulp in a while, but I think these changes are good barring any input from the project on renaming parameters.

sysadmind commented:
shipit

coreaut commented:
@sysadmind I think you need to answer in the referenced issue for someone to check and approve it?

coreaut commented:
reopen because of stale_ci label.

Commit:
Updated version_added

Updated version_added info from 2.9 to 2.10

sysadmind commented:
I think you probably need to rebase this for the CI tests to pass. [A reference on rebase ](https://git-scm.com/book/en/v2/Git-Branching-Rebasing)

sysadmind commented:
shipit

coreaut commented:
bot_broken

thanks for the help, hope I'm not wrong but I think the bot is broken? It ignores the shipit command. Thanks for the help Ansible staff.

jamescassell commented:
I'd drop version_added entirely, then

shipit

I think this should be backported since it's a bugfix and the module hasn't worked properly since 2.8.0 w/o this fix.

Commit:
Update pulp_repo.py

removed version_added for renamed params

tremble commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -561,9 +561,9 @@ def main():
-    importer_ssl_client_cert = module.params['client_cert']
-    importer_ssl_client_key = module.params['client_key']
+    importer_ssl_ca_cert = module.params['feed_ca_cert']
+    importer_ssl_client_cert = module.params['feed_client_cert']
```
```suggestion
    importer_ssl_client_cert = module.params['feed_client_cert']
    if importer_ssl_client_cert is None and module.params['client_cert'] is not None:
      importer_ssl_client_cert = module.params['client_cert']
      module.deprecate("In Ansible 2.10 a new `feed_client_cert` option was added.  Until Ansible 2.14 the default value will come from the client_cert option", version="2.14")
```

coreaut replied:
here's also a accidentally peer in the deprecate :)

tremble commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -535,9 +535,9 @@ def main():
-        ca_cert=dict(aliases=['importer_ssl_ca_cert']),
-        client_cert=dict(aliases=['importer_ssl_client_cert']),
-        client_key=dict(aliases=['importer_ssl_client_key']),
+        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert']),
```
```suggestion
        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert','ca_cert']),
```

tremble commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -48,13 +48,13 @@
       - CA certificate string used to validate the feed source SSL certificate.
         This can be the file content or the path to the file.
     type: str
     aliases: [ importer_ssl_ca_cert ]
```
```suggestion
    aliases: [ importer_ssl_ca_cert, ca_cert ]
```

tremble commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -64,7 +64,7 @@
     type: str
     aliases: [ importer_ssl_client_cert ]
-  client_key:
+  feed_client_key:
```
```suggestion
  feed_client_key:
    version_added: 2.10
```

tremble commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -48,13 +48,13 @@
     type: str
     aliases: [ importer_ssl_ca_cert ]
-  client_cert:
+  feed_client_cert:
```
```suggestion
  feed_client_cert:
    version_added: 2.10
```

tremble reviewed (changes requested):
What I was describing on IRC...

You currently use ``argument_spec = url_argument_spec()``

This will automatically (re-) add client_cert and client_key to your argument_spec and the doc fragment will add them to the documentation.

My suggestion is as follows (changes beyond what you have today):

- add ca_cert as an alias to peer_ca_cert
- add the ``version_added: 2.10`` entries to the documentation
- For now, set the defaults based on the client_cert/client_key values

coreaut commented:
> What I was describing on IRC...
> 
> You currently use `argument_spec = url_argument_spec()`
> 
> This will automatically (re-) add client_cert and client_key to your argument_spec and the doc fragment will add them to the documentation.
> 
> My suggestion is as follows (changes beyond what you have today):
> 
>     * add ca_cert as an alias to peer_ca_cert
> 
>     * add the `version_added: 2.10` entries to the documentation
> 
>     * For now, set the defaults based on the client_cert/client_key values

You mean feed_ca_cert not peer?
I did not know that client_cert and client_key get auto added. But that should be ok, one can configure pulp also to use tls client auth and not http basic auth.

I think however client_cert should not go into importer_ssl_client_cert if set.
Because it could override the certs one wants pulp to use to sync remote repos.

So a scenario like this would work as intended:
ansible -> tls client auth -> pulp server -> tls client auth -> remote repo

tremble commented:
Yes, sorry I meant feed_ca_peer

> I think however client_cert should not go into importer_ssl_client_cert if set.

1. If you set feed_client_cert, this takes precedence.  Always.
2. If you don't set feed_client_cert, but do set client_cert, you get the same behaviour as 2.9
3. If you set feed_client_cert to "" then if won't default to client_cert and the various if clauses will ignore it

coreaut commented:
@tremble thanks for your time and help here. 
Can you change your suggestion to fix the typo about peer in the deprecation warning?
I'm not sure if you got a message about my comment on that one.

tremble commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -561,9 +561,9 @@ def main():
-    importer_ssl_client_key = module.params['client_key']
+    importer_ssl_ca_cert = module.params['feed_ca_cert']
+    importer_ssl_client_cert = module.params['feed_client_cert']
+    importer_ssl_client_key = module.params['feed_client_key']
```
```suggestion
    importer_ssl_client_key = module.params['feed_client_key']
    if importer_ssl_client_key is None and module.params['client_key'] is not None:
      importer_ssl_client_key = module.params['client_key']
      module.deprecate("In Ansible 2.10 a new `feed_client_key` option was added.  Until Ansible 2.14 the default value will come from the client_key option", version="2.14")
```

Commit:
Apply suggestions from code review

added ca_cert alias and 'version_added'

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

Commit:
Apply suggestions from code review

added old behavior for client_cert and client_key which will deprecate in 2.14

Co-Authored-By: Mark Chappell <mchappel@redhat.com>

coreaut reviewed (commented):
keep ca_cert as alias and incorporate old behavior until it deprecates in 2.14, also include added_version info.

Fine by me and thanks!

coreaut commented:
@tremble about the deprecation, before the 2.14 release I have to submit again to remove the then deprecated behavior?

tremble commented:
As soon as 2.13 and 2.14 branch you would be able to do this.  I *believe* this should be in around 18 months time...

Commit:
Update pulp_repo.py

fix for linting error ansibot is complaining

coreaut commented:
@sysadmind @tremble @jamescassell @felixfontein 
I humbly request shipits after you have checked it, thanks.

It can only be backported after the fix in devel is out. 
[https://docs.ansible.com/ansible/latest/community/development_process.html#backporting-merged-prs](url)

Commit:
added changelog fragment for 59522

sysadmind commented:
shipit

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -535,9 +537,9 @@ def main():
-        ca_cert=dict(aliases=['importer_ssl_ca_cert']),
-        client_cert=dict(aliases=['importer_ssl_client_cert']),
-        client_key=dict(aliases=['importer_ssl_client_key']),
+        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert', 'ca_cert']),
```
This is not a good idea, since `ca_cert` is already a module option. (Due to `url_argument_spec()` above.)

Either do
```suggestion
        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert'']),
```
and make sure that `feed_ca_cert` is set to `ca_cert`'s value if not set, or name the option `ca_cert` and add `feed_ca_cert` and `importer_ssl_ca_cert` as aliases. (Then there's no way to separate these options.)

tremble replied:
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L1410

I may have missed it, I just don't see it...

coreaut replied:
@felixfontein I also don't see it being used by url_argument_spec()?

felixfontein replied:
Hmm, I was pretty sure it was in there. Apparently, I'm wrong. Sorry!

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -64,7 +65,8 @@
-  client_key:
+  feed_client_key:
+    version_added: "2.10"
     description:
```
You should mention that the default value is taken from `client_key` if that's specified, and that this will change in Ansible 2.14.

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -48,13 +48,14 @@
+    aliases: [ importer_ssl_ca_cert, ca_cert ]
+  feed_client_cert:
+    version_added: "2.10"
     description:
```
You should mention that the default value is taken from `client_cert` if that's specified, and that this will change in Ansible 2.14.

coreaut replied:
Thanks that's  a good idea. Added a commit for this.

felixfontein commented on changelogs/fragments/59522-renamed-module-tls-client-auth-params-to-avoid-overlaping-with-fetch_url.yaml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - pulp_repo - fix for module shadowing url_argument_spec params (https://github.com/ansible/ansible/issues/59513)
```
That doesn't help anyone understanding what happened without knowing some Ansible internals. How about something like:
```suggestion
  - pulp_repo - the ``client_cert`` and ``client_key`` options were used for both requests to pulp.io and for the repo to sync with, resulting in errors when they were used. Use the new options ``feed_client_cert`` and ``feed_client_key`` for client certificates that should only be used for repo synchronisation, and not for communication with pulp.io. (https://github.com/ansible/ansible/issues/59513)
```

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -561,9 +563,15 @@ def main():
+    importer_ssl_client_cert = module.params['feed_client_cert']
+    if importer_ssl_client_cert is None and module.params['client_cert'] is not None:
+        importer_ssl_client_cert = module.params['client_cert']
+        module.deprecate("In Ansible 2.10 `feed_client_cert` option was added. Until 2.14 the default value will come from client_cert option", version="2.14")
```
How about:
```suggestion
        module.deprecate("To specify client certificates to be used with the repo to sync, and not for communication with pulp.io, use the new options `feed_client_cert` and `feed_client_key` (available since Ansible 2.10). Until Ansible 2.14, the default value for `feed_client_cert` will be taken from `client_cert` if only the latter is specified", version="2.14")
```
(And something similar below.)

felixfontein commented:
If this should be backported, explicitly referencing Ansible 2.10 is not a good idea. That would be wrong, then. I'm not sure how to proceed here, though. Maybe it's best to merge it with 2.10 mentioned, change it to 2.9.x in the backport to `stable-2.9` (whatever the next 2.9.x version will be), and once that's merged, add a PR to `devel` which updates the version numbers accordingly.

Commit:
Apply suggestions from code review

more informative depreciation warning and changelog fragment

Co-Authored-By: Felix Fontein <felix@fontein.de>

Commit:
added mention for changes in client_key and client_cert behavior

coreaut commented:
> If this should be backported, explicitly referencing Ansible 2.10 is not a good idea. That would be wrong, then. I'm not sure how to proceed here, though. Maybe it's best to merge it with 2.10 mentioned, change it to 2.9.x in the backport to `stable-2.9` (whatever the next 2.9.x version will be), and once that's merged, add a PR to `devel` which updates the version numbers accordingly.

I will go with that and change it accordingly later if the backport gets accepted.

Commit:
fixed too long line (linting)

Commit:
deprecated ca_cert alias to have consistent module params in Ansible 2.14

Commit:
fixed indentation for deprecation warning

Commit:
changed deprecated alias handling to argument_spec

Commit:
moved deprecated_aliases insied argument dict, thanks tremble

coreaut commented:
If no one has further suggestions I'm ready with this PR :)

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -48,27 +48,34 @@
         certificate. The specified file may be the certificate itself or a
         single file containing both the certificate and private key. This can be
         the file content or the path to the file.
+        If not specified the default value will come from client_cert. Which will
```
I would add this to a new paragraph, so it will be easier to read (and harder to miss).
```suggestion
      - If not specified the default value will come from client_cert. Which will
```

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -48,27 +48,34 @@
       - Private key to the certificate specified in I(importer_ssl_client_cert),
         assuming it is not included in the certificate file itself. This can be
         the file content or the path to the file.
+        If not specified the default value will come from client_key. Which will
```
Same here:
```suggestion
      - If not specified the default value will come from client_key. Which will
```

coreaut replied:
yes that's a good idea, thanks

felixfontein commented on lib/ansible/modules/packaging/os/pulp_repo.py:
```diff
@@ -535,9 +542,9 @@ def main():
-        ca_cert=dict(aliases=['importer_ssl_ca_cert']),
-        client_cert=dict(aliases=['importer_ssl_client_cert']),
-        client_key=dict(aliases=['importer_ssl_client_key']),
+        feed_ca_cert=dict(aliases=['importer_ssl_ca_cert', 'ca_cert'], deprecated_aliases=[dict(name='ca_cert', version='2.14')]),
```
Cool, I didn't knew we had `deprecated_aliases`! Thanks @tremble!

felixfontein reviewed (commented):
Some details, besides that LGTM!

Commit:
suggestions from felixfontein

Move doc info about client_cert and client_key into its own paragraph

Co-Authored-By: Felix Fontein <felix@fontein.de>

felixfontein reviewed (approved):
shipit

felixfontein commented:
@sysadmind @tremble still happy with this PR? If yes, I'll merge.

tremble commented:
@felixfontein LGTM

felixfontein commented:
If nobody complains I'll merge it on Monday, create a backport and put it on the project meeting agenda for Tuesday to see what the core team thinks about backporting this.

felixfontein commented:
@coreaut thanks for fixing this!
@sysadmind @jamescassell @tremble thanks for reviewing this!
